### ISSUE TITLE
Data Races

### ISSUE THREAD

**Issue by infogulch** [yarf-framework/yarf#12]
yarf appears to have several data races, some of which seem to be part of the framework's API.

Note the documentation of [`http.Serve`](https://golang.org/pkg/net/http/#Server.Serve) (Which is called by [`http.ListenAndServe`](https://golang.org/src/net/http/server.go?s=55886:55927#L1858) among others.)(Emphasis mine):

> Serve accepts incoming connections on the Listener l, **creating a new service goroutine for each**. The service goroutines read requests and then call srv.Handler to reply to them.

I created a [new test](https://github.com/infogulch/yarf/commit/d3dd6129a46ba8b4dc34ea4d873e94b33b508d6d) on my fork that can surfaces data races in `yarf.ServeHTTP`, simulating the way Server behaves by spawning many goroutines to handle many requests. The output of `go test -race` with this test is at the bottom of this issue. The post on the Go Blog, [Intoducing the Go Race Detector](https://blog.golang.org/race-detector) may be an insightful read.

1. The first is a race on the cache itself. The first goroutine is reading from it at the same time as another is writing to it. As a stop-gap it could use a [`sync.RWMutex`](https://golang.org/pkg/sync/#RWMutex) to serialize access to it. But since the current cache implementation is quite simplified, it might be better off removed from yarf completely and implemented at a different layer.
2. The second race is on `routeGroup.lastMatch`. This could result in one goroutine overwriting `lastMatch` between another goroutine's write of `lastMatch` and call of `Dispatch`, causing the same Route to be dispatched to twice with two or more completely different contexts, and the other routes ignored completely. To fix this, I would put the last match in a new Dispatchee field of Context, and remove it from RouteGroup.
3. The third race is on a Route's Context field (via embedded RequestContext). This means that two goroutines could call `SetContext` before either calls `Dispatch`, resulting in both routes getting the same context. Since this is part of the yarf's API, I doubt this could be fixed without breaking things. One change would be to change `ResourceHandler` http method handlers to take `*Context` as an argument instead of being passed underneath in the struct with `RequestContext`.

This issue doesn't come with a pull request just because applying these changes would alter the API and I'd like your feedback first.

```
==================
WARNING: DATA RACE
Read by goroutine 39:
  runtime.mapaccess2_faststr()
      C:/sdk/Go/src/runtime/hashmap_fast.go:281 +0x0
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:83 +0x23a
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:131 +0xc1

Previous write by goroutine 38:
  runtime.mapassign1()
      C:/sdk/Go/src/runtime/hashmap.go:411 +0x0
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:98 +0x564
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:131 +0xc1

Goroutine 39 (running) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3

Goroutine 38 (finished) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3
==================
==================
WARNING: DATA RACE
Read by goroutine 40:
  github.com/yarf-framework/yarf.(*routeGroup).Dispatch()
      go/src/github.com/yarf-framework/yarf/router.go:158 +0x60
  github.com/yarf-framework/yarf.(*yarf).dispatch()
      go/src/github.com/yarf-framework/yarf/yarf.go:135 +0x1bc
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:88 +0x30d
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:131 +0xc1

Previous write by goroutine 38:
  github.com/yarf-framework/yarf.(*routeGroup).Match()
      go/src/github.com/yarf-framework/yarf/router.go:146 +0x2a5
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:95 +0x459
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:131 +0xc1

Goroutine 40 (running) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3

Goroutine 38 (finished) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3
==================
==================
WARNING: DATA RACE
Write by goroutine 39:
  github.com/yarf-framework/yarf.(*MockResource).SetContext()
      <autogenerated>:83 +0x64
  github.com/yarf-framework/yarf.(*route).Dispatch()
      go/src/github.com/yarf-framework/yarf/router.go:63 +0x81
  github.com/yarf-framework/yarf.(*routeGroup).Dispatch()
      go/src/github.com/yarf-framework/yarf/router.go:175 +0x2e5
  github.com/yarf-framework/yarf.(*yarf).dispatch()
      go/src/github.com/yarf-framework/yarf/yarf.go:135 +0x1bc
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:88 +0x30d
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:131 +0xc1

Previous write by goroutine 38:
  github.com/yarf-framework/yarf.(*MockResource).SetContext()
      <autogenerated>:83 +0x64
  github.com/yarf-framework/yarf.(*route).Dispatch()
      go/src/github.com/yarf-framework/yarf/router.go:63 +0x81
  github.com/yarf-framework/yarf.(*routeGroup).Dispatch()
      go/src/github.com/yarf-framework/yarf/router.go:175 +0x2e5
  github.com/yarf-framework/yarf.(*yarf).dispatch()
      go/src/github.com/yarf-framework/yarf/yarf.go:135 +0x1bc
  github.com/yarf-framework/yarf.(*yarf).ServeHTTP()
      go/src/github.com/yarf-framework/yarf/yarf.go:102 +0x599
  github.com/yarf-framework/yarf.TestYarfParallel.func1()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:132 +0x12c

Goroutine 39 (running) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3

Goroutine 38 (finished) created at:
  github.com/yarf-framework/yarf.TestYarfParallel()
      go/src/github.com/yarf-framework/yarf/yarf_test.go:133 +0x4f9
  testing.tRunner()
      C:/sdk/Go/src/testing/testing.go:456 +0xe3
==================
```

**Comment by leonelquinteros** [comment]
Thanks for finding this. It's a big one! 

1. We'll figure out something, as you said, being super simple we can think on another way to handle this, i'd like to keep the route cache as it's a nice performance improver. 

2. I'm already working on a fix for this one based on your proposal. I'll let you know. 

3. The idea of not passing the Context around was to simplify development and it's something i'd like to keep (as is mentioned on the Readme file. But now i see this, i'm not sure that's possible, but let's think about it. 

Proposals for the cache layer are welcome. 

If you have any idea on how to handle the Context object without having to pass it around methods please send it over. I guess we may implement a custom ResponseWriter and insert the Context there (not sure if that will work though) but also it's a bit weir to have the Context in the Response object. 

Lets think about that, this is getting fun! 

Thanks again for all your contributions. I hope you like the framework and let me know if you are thinking on using it on some project.

**Comment by leonelquinteros** [comment]
Fixed on https://github.com/yarf-framework/yarf/commit/ddd49168504cdf8be6221dadb4417918ede9d0f4 

1. Implemented the sync.RWMutex as you suggested. Found the implementation very simple and the performance benefits are still there. 

2. Also following your proposal, moved the route group last match to the Context object. 

3. After trying several approaches, I found that the most convenient solution was your proposal to set the Context object as a parameter to all Resource methods. This removed some code overhead about the context (i.e. RequestContext object), and changed the API (as well as the version number). 

Check that i did several other commits after this one that fixes some errors and improves some other things along with the update on the docs to match the new API. 
Examples are also updated. 

I wrote a test for the race conditions you mentioned, a bit different from yours to replicate the behaviour of the http server without using the sync package. The outcome should be the same and i've also checked your test to confirm that both are passing OK. 

Thanks again for bringing this up!
